rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -865,63 +975,123 @@ void main_window::HandlePupInstallation(QString file_path)
+		gui_log.error("%s", pup.get_formatted_error());
 		critical(tr("Firmware installation failed: The provided file is incomplete. Try redownloading it."));
 		return;
 	}
-
-	pup_object pup(pup_f);
-	if (!pup)
+	case pup_error::header_file_count:
+	case pup_error::file_entries:
+	default:
 	{
-		gui_log.error("Error while installing firmware: PUP file is invalid.");
+		std::string error = "Error while installing firmware: PUP file is invalid.";
+
+		if (!pup.get_formatted_error().empty())
+		{
+			fmt::append(error, "\n%s", pup.get_formatted_error());
+		}
+
+		gui_log.error("%s", error);
 		critical(tr("Firmware installation failed: The provided file is corrupted."));
 		return;
 	}
-
-	if (!pup.validate_hashes())
+	case pup_error::hash_mismatch:
 	{
 		gui_log.error("Error while installing firmware: Hash check failed.");
 		critical(tr("Firmware installation failed: The provided file's contents are corrupted."));
 		return;
 	}
+	case pup_error::ok: break;
+	}
 
 	fs::file update_files_f = pup.get_file(0x300);
+
+	if (!update_files_f)
+	{
+		gui_log.error("Error while installing firmware: Couldn't find installation packages database.");
+		critical(tr("Firmware installation failed: The provided file's contents are corrupted."));
+		return;
+	}
+
 	tar_object update_files(update_files_f);
+
+	if (!dir_path.isEmpty())
+	{
+		// Extract only mode, extract direct TAR entries to a user directory
+
+		if (!vfs::mount("/pup_extract", sstr(dir_path) + '/'))
+		{
+			gui_log.error("Error while extracting firmware: Failed to mount '%s'", sstr(dir_path));
+			critical(tr("Firmware extraction failed: VFS mounting failed."));
+			return;	
+		}
+
+		if (!update_files.extract("/pup_extract"))
+		{
+			gui_log.error("Error while installing firmware: TAR contents are invalid.");
+			critical(tr("Firmware installation failed: Firmware contents could not be extracted."));
+		}
+
+		gui_log.success("Extracted PUP file to %s!", sstr(dir_path));
+		return;
+	}
+
+	// In regular installation we select specfic entries from the main TAR which are prefixed with "dev_flash_"
+	// Those entries are TAR as well, we extract their packed files from them and that's what installed in /dev_flash
+
 	auto update_filenames = update_files.get_filenames();
 
 	update_filenames.erase(std::remove_if(
 		update_filenames.begin(), update_filenames.end(), [](std::string s) { return s.find("dev_flash_") == umax; }),
 		update_filenames.end());
 
-	std::string version_string = pup.get_file(0x100).to_string();
+	if (update_filenames.empty())
+	{
+		gui_log.error("Error while installing firmware: No dev_flash_* packages were found.");
+		critical(tr("Firmware installation failed: The provided file's contents are corrupted."));
+		return;
+	}
+
+	static constexpr std::string_view cur_version = "4.87";
+
+	std::string version_string = [](fs::file version){ return version ? version.to_string() : std::string{}; }(pup.get_file(0x100));
```

Comment:
this is hard to read.
just do:
```suggestion
	std::string version_string;
	if (fs::file version_file = pup.get_file(0x100))
		version_string = version_file.to_string();
```